### ISSUE TITLE
Option To Load Favicons For Imported Bookmarks

### ISSUE THREAD

**Issue by opticyclic** [Emdek/otter#667]
After importing bookmarks the favicons are some default icon until you visit that page.
It would be nice to have the option to load the favicons without visiting all the pages.
Some people have huge amounts of bookmarks so it would be good to have it as a background process that didn't prevent you browsing whilst it ran.

**Comment by Emdek** [comment]
We need to create generic way to store them first, separate from backends.
It's hard to decide which format would be best for them, for sure not the way how it was done in Opera 12 (thousands of small files in single directory).

**Comment by cofi89** [comment]
Visiting a bookmarked page doesn't always change/load the favicon in the bookmarks toolbar. It just keeps using the default icon, although the favicon is always shown on the tab and adress bar ( if the site has one ofc ). 
And it seems to go like that completelly random, for some b.marks/sites it gets properly loaded, for some it doesn't.
Bookmarks are imported from Firefox.
I don't know if it's related to #820 but I've been seeing this behavior since b.marks toolbar became available. 
Screenshot: http://en.zimagez.com/zimage/screenshot-10052015-123423am.php 
Otter report: https://paste.debian.net/314433/ 

Cheers! :smiley:

**Comment by Emdek** [comment]
@cofi89, yes, it is caused by the issue to be fixed by #820, this one is meant for tracking progress on importing favicons together with bookmarks.
It seems that in case of bookmarks it could be a good idea to allow to store them with bookmarks (embed them into *bookmarks.xbel*), that would allow to set custom icons etc.

**Comment by cofi89** [comment]
Yeah, but I've had more "luck", so to speak, getting the favicons to show for Otter (re)created bookmarks then for the imported ones. Also, wasn't sure if it certanly is related to #820, so I posted here.

@xbel/embeding:
Sorry for my ignorance, but if I'm understanding correctly, embedding favicons in bmarks.xbel would be a "fix" for bookmarks, but not for history/tabs/lists/etc (820). That is by "impoving favicon storage" you tought that it'd be the best to use a single, central storage spot/db?  
If so, (with my very limited knowledge of the subject) I guess that it comes down to what is easier to implement/maintain or works better...
To me, it seems that keeping them separated might need less coding/work/maintainance since xbel is a standard (xml)? :smiley: 

On the side note, Konqueror and Aurora both use it as a default way of storing bookmarks and Konqueror has the abillity to import/refresh (all) favicons, so it might be a good place to look at/borrow ideas ( if needed/helps ofc )...

Also, +1 for ability to set custom icons.

**Comment by Emdek** [comment]
@cofi89, yes, embedding would help mostly for bookmarks themselves and it is requirement for setting custom icons. We could could consider use of these pixmaps as fallback in history if icon won't be available in our central storage (#820).
Basically central storage is meant for use only by history and related stuff (search services embed own icons in their definitions, according to OpenSearch standard).

While XBEL is a standard it doesn't have a spec for embedding icons, although it has a way for storing custom metadata, which we can use for that purpose.
